devcareer/code-jammers-backend
# #174904959 Build APIs for Get country, Get countries, Edit countries and Delete countries

## What does this PR do ##
-  enables user to get one country with specific id 
-  enables user to get all countries
- enables admin to edit an existing country in the database
- enables admin to delete a country from the database

### Description of tasks to be completed? ###
- create get single country route
- create get multiple countries routes
- create get single country controller
- create get multiple countries controller
- create a delete country controller
- create a delete country route
- create an edit country controller
- create an edit country route
- write tests for the controllers and routes

### How should this be manually tested/checked? ###
- Clone the PR
- Checkout to the branch `ft-user-should-get-countries-174904959`
- Install the dependencies using `npm install`
- Create a `.env file` and define the following as follows:
   - `DEV_DATABASE_URL = <development database url>`
   - `TEST_DATABASE_URL = <test database url>`
   - `DATABASE_URL = <production database url>`
- run ```npm run migrate``` to create the database tables
- run ```npm run seed``` to seed the database
- Start the server by running ```npm run start-dev```
- make a GET request to the url endpoint http://localhost:3000/api/v1/country/2e11e4a9-441b-4426-9521-39adc64ccfad
  it should return the country with that id for you
- make GET a request to url endpoint http://localhost:3000/api/v1/countries
 it should return all the countries in the database
### sign in and copy the token paste it in the authorization header to test the following admin routes
- make a PATCH request with this object in the request body {"population":200000} to the url endpoint http://localhost:3000/api/v1/admin/country/2e11e4a9-441b-4426-9521-39adc64ccfad
it should update the country's population
- make a DELETE request to the url endpoint http://localhost:3000/api/v1/admin/country/2e11e4a9-441b-4426-9521-39adc64ccfad
it should delete the country with specified id

### Link to architecture
[architecture](https://docs.google.com/document/d/1z6eZzaxyaP68EFPCy_sBPsSUSrAeFNiIfl5AGsxM_aQ/edit)

### What are the relevant pivotal tracker stories?
[#174904959](https://www.pivotaltracker.com/story/show/174904959)
